1346 Microfrontend for meldekort

Henreich opened the pull request:
**Data**
Satt opp endepunkt for å gi informasjon om hvor mange meldekort brukeren kan fylle ut og eventuelt når neste meldekort kan sendes inn

**Aktivering og inaktivering av microfrontend:**
Satt det opp slik at microfrontend aktiveres dersom man har en meldeperiode hvor det er minst en dag som har gir_rett=true, og til og med datoen er innenfor den siste måneden. 

Tilsvarende deaktiveres microfrontenden når brukers siste meldeperiode har minst en dag som har gir_rett=true og det er en måned siden tilOgMed

https://trello.com/c/Dj1Ji9PR/1346-microfrontend-til-minside-for-meldekort

Henreich committed:
1346 Informer bruker om når de kan fylle ut meldekort i microfrontend

Henreich committed:
1346 Informer bruker om når de kan fylle ut meldekort i microfrontend

Henreich committed:
1346 Routing og URL

Henreich committed:
1346 Deaktiver microfrontend for brukere som ikke har hatt en meldeperiode med rett de siste 6 måneder

Henreich committed:
1346 Tester for deaktivering av microfrontend

Testene er ikke helt nøye på dato fordi bruk av nå() og perioder gjør at til og med dato flytter på seg, basert på hvilken dag i uka det er. Dog er ikke dette tester for å sjekke periode

Henreich committed:
1346 Aktiver microfrontend i en egen jobb

Henreich commented on src/main/kotlin/no/nav/tiltakspenger/meldekort/repository/SakPostgresRepo.kt:
```diff
@@ -27,20 +29,23 @@ class SakPostgresRepo(
                     "fnr" to sak.fnr.verdi,
                     "arena_meldekort_status" to sak.arenaMeldekortStatus.tilDb(),
                     "har_soknad_under_behandling" to sak.harSoknadUnderBehandling,
+                    "microfrontend_inaktivert" to sak.erMicrofrontendInaktivert,
```
Hadde misforstått bruken av lagre metoden her, ser i loggene i dev at man forsøker å lage en ny her.. ikke lagre en eksisterende behandling. Må ta en runde til :D

anders-nom commented on src/main/kotlin/no/nav/tiltakspenger/meldekort/service/MeldekortService.kt:
```diff
@@ -119,4 +121,18 @@ class MeldekortService(
+            // Siden vi filtrerer på IKKE_KLAR så er klarTilInnsendingDag aldri null her
+            .klarTilInnsendingDag!!
+
+        return Pair(antallMeldekortBrukerKanfylleut, nesteMuligeInnsending)
```
Lurer på om dette kan være verdt å optimalisere litt, kanskje heller tilpasse sql-spørringen til å hente ut antall meldekort og neste innsending, istedenfor å hente ut alt fra db hver gang.

Henreich replied:
jeg begynte med å gjøre filtreringen i spørringen, men problemet da er at jeg trenger egentlig samme logikk som klarTilInnsending og status feltet utledes av, bare at jeg må duplisere logikken under i sql-spørringen også. Det er jo litt uheldig med at samme ting utledes to steder 😅 
```kotlin
    val status: MeldekortStatus = when {
        mottatt != null -> MeldekortStatus.INNSENDT
        deaktivert != null -> MeldekortStatus.DEAKTIVERT
        !periode.tilOgMed.isAfter(senesteTilOgMedDatoForInnsending()) -> MeldekortStatus.KAN_UTFYLLES
        else -> MeldekortStatus.IKKE_KLAR
    }

    val klarTilInnsending: Boolean by lazy { status == MeldekortStatus.KAN_UTFYLLES }
```

anders-nom replied:
Mm, er ikke sikker på beste løsning her egentlig. :sweat_smile:

Går uansett fint å merge denne nå, lurer bare på om det ikke vil skalere så bra på sikt når vi får mange brukere med lang meldekort-historikk.

Henreich replied:
ettersom jeg måtte fikse at det ble kasta exception så la jeg inn tilsvarende logikk i spørringa liksågodt

anders-nom commented on src/main/kotlin/no/nav/tiltakspenger/meldekort/domene/Sak.kt:
```diff
@@ -14,6 +14,7 @@ data class Sak(
     val meldeperioder: List<Meldeperiode>,
     val arenaMeldekortStatus: ArenaMeldekortStatus,
     val harSoknadUnderBehandling: Boolean,
+    val erMicrofrontendInaktivert: Boolean,
```
Husk å oppdatere erLik-funksjonen til å ignorere dette flagget :sweat_smile:

anders-nom commented on src/main/kotlin/no/nav/tiltakspenger/meldekort/repository/SakPostgresRepo.kt:
```diff
@@ -57,12 +59,14 @@ class SakPostgresRepo(
                     "id" to sak.id.toString(),
                     "fnr" to sak.fnr.verdi,
                     "har_soknad_under_behandling" to sak.harSoknadUnderBehandling,
+                    "microfrontend_inaktivert" to sak.erMicrofrontendInaktivert,
```
Når en ny versjon av saken fra sbh-api oppdateres, så vil dette flagget overskrives med default-verdien. Kanskje det hadde vært ryddigst med en separat funksjon for å oppdatere denne?

Henreich replied:
ah, ser jeg hadde misforstått her. Good catch

Henreich committed:
1346 Egen metode for å oppdatere flagget for microfrontend og ignorerer flagget for sammenligning (erLik)

Henreich committed:
1346 Optimaliserer query til å bare returnere meldekort bruker kan fylle ut

Henreich committed:
1346 Sjekk negative tilfeller i aktivering/inaktivering for å unngå at flagget endres frem og tilbake

Henreich committed:
1346 Flagget brukes nå som en status

Henreich committed:
1346 Endrer offset for når microfrontend inaktiveres fra 6 til 1 måned

Det kom fram i diskusjoner at 6 måneder er kanskje litt overdrevet. Microfrontendene tar mye plass og oppmerksomhet på din side, så da burde det være synlig litt kortere

Henreich committed:
1346 Skrur av jobbene for microfrontend i prod frem til vi har formell godkjenning av Team Min Side

Min side teamet ønsker å teste det i dev før vi får tilgang til kafka-topicet for microfrontend i prod
